spinnaker-monitoring-daemon/spinnaker-monitoring/datadog_service.py

```diff
@@ -20,17 +20,168 @@
+                self.__lookup_in_agent_config('hostname') or
+                ''
+              ),
+    }
+
+    required_options.update(nonessential_options)
+
+    return required_options
+
+  def __contains_value(self, key, dictionary):
+    return key in dictionary and dictionary[key] is not None
+
+  def __resolve_value(self, identifier, key, required=True):
+    """
+    Resolve to a value set in either an environment variable,
+    in the outer section of spinnaker-monitoring.yml, within the
+    datadog field of spinnaker-monitoring.yml, in the datadog
+    agent configuration file, all in that order, or error out.
+    Whichever source provides a non-None value first wins.
+
+    Order of lookup is as as follows:
+
+    1. Check for string 'DATADOG_%s' % key.toupper() as
+       environment variable.
+    2. Check for string 'datadog_%s' % key in the
+       spinnaker-monitoring.yaml.
+    3. Check for key as just '%s' % key in the datadog
+       section in the spinnaker-monitoring.yaml
+    4. Check for key as just '%s' % key in the Datadog
+       agent configuration.
+
+    If, after this, `required` is False, don't throw an error
+    but return None instead.
+
+    Parameter:
+      identifier: String. If value could not be found,
+                  use this parameter in the error message
+                  to refer to what you were actually trying to find.
+      key: String. Look for this key.
+      required: Bool. If the value is not required and
+                could not be found, return None.
+    """
+
+    environment_key = 'DATADOG_{0}'.format(key.upper())
+    yaml_key = 'datadog_{0}'.format(key)
+
+    # check if value is set as an environment variable
+    if self.__contains_value(environment_key, os.environ):
+      return os.environ[environment_key]
+
+    # check if value is set in spinnaker-monitoring.yml as a high-level key
+    if self.__contains_value(yaml_key, self.options):
+      return self.options[yaml_key]
+
+    # check if value is set in spinnaker-monitoring.yml as a key under the datadog
+    # section.
+    if self.__contains_value(key, self.options['datadog']):
+      return self.options['datadog'][key]
+
+    # check finally if value is set in Datadog agent configuration.
+    configuration_value = self.__lookup_in_agent_config(key)
+    if configuration_value is not None:
+      return configuration_value
+
+    if not required:
+      return None
+
+    raise ValueError('{0} could not be found as environment '
+                     'variable {1}, or read from spinnaker-monitoring.yml file as '
+                     '{2} as an outer value, or found as {3} under section datadog '
+                     'and could not be read from datadog agent configuration using'
+                     'key {3}'
+                     .format(identifier, environment_key, yaml_key, key)
+                     )
+
+  def __convert_to_list_of_strings(self, argument, default=[]):
```

Comment:
default=None

It is generally bad to have default objects because the same default value is used for every invocation. Thus when the second caller uses the default value, they will get the same reference as the first caller, thus the values are the same object even though the callers might be completely unrelated.

While this is private and probably "one use only", I think it is still dangerous from a maintainability standpoint. And might affect tests that violate that "one use only" production assumption not actually testing what you think they are.